src/usr.bin/calendar/calendar.1
.\"	$MirOS: src/usr.bin/calendar/calendar.1,v 1.4 2006/11/17 02:48:09 tg Exp $
.\"	$OpenBSD: calendar.1,v 1.29 2005/03/22 10:57:27 xsa Exp $
.\"
.\" Copyright (c) 1989, 1990, 1993
.\"     The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)calendar.1  8.1 (Berkeley) 6/29/93
.\"
.Dd November 17, 2006
.Dt CALENDAR 1
.Os
.Sh NAME
.Nm calendar
.Nd reminder service
.Sh SYNOPSIS
.Nm calendar
.Op Fl ab
.Op Fl A Ar num
.Op Fl B Ar num
.Op Fl f Ar calendarfile
.Op Fl t Ar [[[cc]yy][mm]]dd
.Sh DESCRIPTION
The
.Nm
utility checks the current directory or the directory specified by the
.Ev CALENDAR_DIR
environment variable for a file named
.Pa calendar
and displays lines that begin with either today's date
or tomorrow's.
On Fridays, events on Friday through Monday are displayed.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl A Ar num
Print lines from today and next
.Ar num
days (forward, future).
.It Fl a
Process the
.Dq calendar
files of all users and mail the results
to them.
This requires superuser privileges.
.It Fl B Ar num
Print lines from today and previous
.Ar num
days (backward, past).
.It Fl b
Enforce special date calculation mode for KOI8 calendars.
.It Fl f Pa calendarfile
Use
.Pa calendarfile
as the default calendar file.
.It Fl t Ar [[[cc]yy][mm]]dd
Act like the specified value is
.Dq today
instead of using the current date.
.El
.Pp
To handle calendars in your national code table you can specify
.Dq LANG=<locale_name>
in the calendar file as early as possible.
To handle national Easter names in the calendars,
.Dq Easter=<national_name>
(for Catholic Easter) or
.Dq Paskha=<national_name>
(for Orthodox Easter) can be used.
.Pp
The
.Dq CALENDAR
variable can be used to specify the style.
Only
.Sq Julian
and
.Sq Gregorian
styles are currently supported.
Use
.Dq CALENDAR=
to return to the default (Gregorian).
.Pp
To enforce special date calculation mode for Cyrillic calendars
you should specify
.Dq LANG=<local_name>
and
.Dq BODUN=<bodun_prefix>
where <local_name> can be ru_RU.KOI8-R, uk_UA.KOI8-U or by_BY.KOI8-B.
.Pp
Other lines should begin with a month and day.
They may be entered in almost any format, either numeric or as character
strings.
If proper locale is set, national months and weekdays
names can be used.
On
.Ox
and
.Mx ,
support for locales is non-existent.
A single asterisk (`*') matches every month.
A day without a month matches that day of every week.
A month without a day matches the first of that month.
Two numbers default to the month followed by the day.
Lines with leading tabs default to the last entered date, allowing
multiple line specifications for a single date.
.Dq Easter
(may be followed by a positive or negative integer) is Easter for this year.
.Dq Paskha
(may be followed by a positive or negative integer) is
Orthodox Easter for this year.
Weekdays may be followed by
.Dq -4
.Li ...
.Dq +5
(aliases last, first, second, third, fourth) for moving events like
.Dq the last Monday in April .
.Pp
By convention, dates followed by an asterisk
.Pq Sq *
are not fixed, i.e., change
from year to year.
.Pp
Day descriptions start after the first <tab> character in the line;
if the line does not contain a <tab> character, it isn't printed out.
If the first character in the line is a <tab> character, it is treated as
the continuation of the previous description.
.Pp
The calendar file is preprocessed by
.Xr cpp 1 ,
allowing the inclusion of shared files such as company holidays or
meetings.
If the shared file is not referenced by a full pathname,
.Xr cpp 1
searches in the current (or home) directory first, and then in the
directory
.Pa /usr/share/calendar .
Empty lines and lines protected by the C commenting syntax
.Pq Li /* ... */
are ignored.
.Pp
Some possible calendar entries
(a \fB\et\fR sequence denotes a <tab> character):
.Bd -unfilled -offset indent
LANG=C
Easter=Ostern

#include <calendar.usholiday>
#include <calendar.birthday>

6/15\fB\et\fRJune 15 (if ambiguous, will default to month/day).
Jun. 15\fB\et\fRJune 15.
15 June\fB\et\fRJune 15.
Thursday\fB\et\fREvery Thursday.
June\fB\et\fREvery June 1st.
15 *\fB\et\fR15th of every month.

May Sun+2\fB\et\fRsecond Sunday in May (Muttertag)
04/SunLast\fB\et\fRlast Sunday in April,
\fB\et\fRsummer time in Europe
Easter\fB\et\fREaster
Ostern-2\fB\et\fRGood Friday (2 days before Easter)
Paskha\fB\et\fROrthodox Easter
.Ed
.Sh FILES
.Bl -tag -width ~/.etc/calendar/calendar -compact
.It Pa calendar
File in current directory.
.It Pa ~/.etc/calendar
Directory in the user's home directory (which
.Nm
changes into, if it exists).
.It Pa ~/.etc/calendar/calendar
File to use if no calendar file exists in the current directory.
.It Pa ~/.etc/calendar/nomail
.Nm
will not send mail if this file exists.
.It Pa calendar.all
International and national calendar files.
.It Pa calendar.birthday
Births and deaths of famous (and not-so-famous) people.
.It Pa calendar.christian
Christian holidays (should be updated yearly by the local system administrator
so that roving holidays are set correctly for the current year).
.It Pa calendar.computer
Days of special significance to computer people.
.It Pa calendar.croatian
Croatian calendar.
.It Pa calendar.fictional
Fantasy and fiction dates (mostly LOTR).
.It Pa calendar.french
French calendar.
.It Pa calendar.german
German calendar.
.It Pa calendar.history
Everything else, mostly U.S. historical events.
.It Pa calendar.holiday
Other holidays (including the not-well-known, obscure, and
.Em really
obscure).
.It Pa calendar.judaic
Jewish holidays (should be updated yearly by the local system administrator
so that roving holidays are set correctly for the current year).
.It Pa calendar.music
Musical events, births, and deaths (strongly oriented toward rock n' roll).
.It Pa calendar.openbsd
.Ox
and
.Mx
related events.
.It Pa calendar.pagan
Pagan holidays, celebrations and festivals.
.It Pa calendar.russian
Russian calendar.
.It Pa calendar.usholiday
U.S. holidays.
.It Pa calendar.world
World wide calendar.
.El
.Sh SEE ALSO
.Xr at 1 ,
.Xr cal 1 ,
.Xr cpp 1 ,
.Xr mail 1 ,
.Xr cron 8
.Sh STANDARDS
The
.Nm
program previously selected lines which had the correct date anywhere
in the line.
This is no longer true: the date is only recognized when it occurs
at the beginning of a line.
.Sh HISTORY
A
.Nm
command appeared in
.At v7 .
.Pp
Since
.Mx 10 ,
if the
.Dv UNICODE
variable is set during compilation,
.Xr iconv 3
is used to convert any input to OPTU-8 output.
.Sh BUGS
.Nm
doesn't handle all Jewish holidays or moon phases.


src/usr.bin/calendar/calendar.c
/**	$MirOS: src/usr.bin/calendar/calendar.c,v 1.2 2005/03/13 18:32:44 tg Exp $ */
/*	$OpenBSD: calendar.c,v 1.23 2004/12/10 15:31:01 mickey Exp $	*/

/*
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__COPYRIGHT("@(#) Copyright (c) 1989, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <err.h>
#include <errno.h>
#include <locale.h>
#include <login_cap.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <tzfile.h>
#include <unistd.h>

#include "pathnames.h"
#include "calendar.h"

__SCCSID("@(#)calendar.c  8.3 (Berkeley) 3/25/94");
__RCSID("$MirOS: src/usr.bin/calendar/calendar.c,v 1.2 2005/03/13 18:32:44 tg Exp $");

char *calendarFile = "calendar";  /* default calendar file */
char *calendarHome = ".etc/calendar"; /* HOME */
char *calendarNoMail = "nomail";  /* don't sent mail if this file exists */

struct passwd *pw;
int doall = 0;
time_t f_time = 0;
int bodun_always = 0;

int f_dayAfter = 0; /* days after current date */
int f_dayBefore = 0; /* days before current date */

struct specialev spev[NUMEV];

void childsig(int);

int
main(int argc, char *argv[])
{
	int ch;
	char *caldir;

#ifndef __MirBSD__
	(void)setlocale(LC_ALL, "");
#endif

	while ((ch = getopt(argc, argv, "abf:t:A:B:-")) != -1)
		switch (ch) {
		case '-':		/* backward contemptible */
		case 'a':
			if (getuid())
				errx(1, "%s", strerror(EPERM));
			doall = 1;
			break;

		case 'b':
			bodun_always++;
			break;

		case 'f': /* other calendar file */
		        calendarFile = optarg;
			break;

		case 't': /* other date, undocumented, for tests */
			if ((f_time = Mktime(optarg)) <= 0)
				errx(1, "specified date is outside allowed range");
			break;

		case 'A': /* days after current date */
			f_dayAfter = atoi(optarg);
			break;

		case 'B': /* days before current date */
			f_dayBefore = atoi(optarg);
			break;

		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (argc)
		usage();

	/* use current time */
	if (f_time <= 0)
	    (void)time(&f_time);

	if (f_dayBefore) {
		/* Move back in time and only look forwards */
		f_dayAfter += f_dayBefore;
		f_time -= SECSPERDAY * f_dayBefore;
		f_dayBefore = 0;
	}
	settime(&f_time);

	if (doall) {
		pid_t kid, deadkid;
		int kidstat, kidreaped, runningkids;
		int acstat;
		struct stat sbuf;
		time_t t;
		unsigned int sleeptime;

		signal(SIGCHLD, childsig);
		runningkids = 0;
		t = time(NULL);
		while ((pw = getpwent()) != NULL) {
			acstat = 0;
			/* Avoid unnecessary forks.  The calendar file is only
			 * opened as the user later; if it can't be opened,
			 * it's no big deal.  Also, get to correct directory.
			 * Note that in an NFS environment root may get EACCES
			 * on a chdir(), in which case we have to fork.  As long as
			 * we can chdir() we can stat(), unless the user is
			 * modifying permissions while this is running.
			 */
			if (chdir(pw->pw_dir)) {
				if (errno == EACCES)
					acstat = 1;
				else
					continue;
			}
			if (stat(calendarFile, &sbuf) != 0) {
				if (chdir(calendarHome)) {
					if (errno == EACCES)
						acstat = 1;
					else
						continue;
				}
				if (stat(calendarNoMail, &sbuf) == 0 ||
				    stat(calendarFile, &sbuf) != 0)
					continue;
			}
			sleeptime = USERTIMEOUT;
			switch ((kid = fork())) {
			case -1:	/* error */
				warn("fork");
				continue;
			case 0:	/* child */
#ifndef __MirBSD__
				(void)setlocale(LC_ALL, "");
#endif
				if (setusercontext(NULL, pw, pw->pw_uid,
				    LOGIN_SETALL ^ LOGIN_SETLOGIN))
					err(1, "unable to set user context (uid %u)",
					    pw->pw_uid);
				if (acstat) {
					if (chdir(pw->pw_dir) ||
					    stat(calendarFile, &sbuf) != 0 ||
					    chdir(calendarHome) ||
					    stat(calendarNoMail, &sbuf) == 0 ||
					    stat(calendarFile, &sbuf) != 0)
						exit(0);
				}
				cal();
				exit(0);
			}
			/* parent: wait a reasonable time, then kill child if
			 * necessary.
			 */
			runningkids++;
			kidreaped = 0;
			do {
				sleeptime = sleep(sleeptime);
				/* Note that there is the possibility, if the sleep
				 * stops early due to some other signal, of the child
				 * terminating and not getting detected during the next
				 * sleep.  In that unlikely worst case, we just sleep
				 * too long for that user.
				 */
				for (;;) {
					deadkid = waitpid(-1, &kidstat, WNOHANG);
					if (deadkid <= 0)
						break;
					runningkids--;
					if (deadkid == kid) {
						kidreaped = 1;
						sleeptime = 0;
					}
				}
			} while (sleeptime);

			if (!kidreaped) {
				/* It doesn't _really_ matter if the kill fails, e.g.
				 * if there's only a zombie now.
				 */
				(void)kill(kid, SIGTERM);
				warnx("uid %u did not finish in time", pw->pw_uid);
			}
			if (time(NULL) - t >= SECSPERDAY)
				errx(2, "'calendar -a' took more than a day; stopped at uid %u",
				    pw->pw_uid);
		}
		for (;;) {
			deadkid = waitpid(-1, &kidstat, WNOHANG);
			if (deadkid <= 0)
				break;
			runningkids--;
		}
		if (runningkids)
			warnx(
"%d child processes still running when 'calendar -a' finished", runningkids);
	}
	else if ((caldir = getenv("CALENDAR_DIR")) != NULL) {
		if(!chdir(caldir))
			cal();
	} else
		cal();

	exit(0);
}


void
usage(void)
{
	(void)fprintf(stderr,
	    "usage: calendar [-ab] [-A num] [-B num] [-f calendarfile] "
	    "[-t [[[cc]yy][mm]]dd]\n");
	exit(1);
}


void
childsig(int signo)
{
}
